### DISCUSSION TITLE
How to handle dynamic client-side pages?

### DISCUSSION THREAD

**Discussion by jeremyhaile** [vercel/next.js#50310] [App Router] [answered]
OK...new to next.js and I jumped right into the app router. However - because I'm pulling in lots of other react components, many of my pages are "use client"  Now that I'm trying to hook those components up to dynamic APIs, I'm running into two really frustrating issue:

1. When I try to build my app, the build is failing with "Error occurred prerendering page"....but I can't understand why it would pre-render a page that is loading data from an API...the data is dynamic. I see that in the old version of next.js prerendering was disabled if you use getInitialProps, but this isn't supported in "app routing" so how do I stop it from pre-rendering dynamic pages.

2. Sort of related...I need to use session data to update an Authorization header for API requests. I don't see any docs for configuring base URLs or dynamic, session-based headers using fetch. So I created a React module to wrap Axios and used interceptors which works but seems to go against everything the Next.JS docs say about using fetch. Is that the right approach or is there a better way?

I really appreciate any support, because I've been banging my head against the wall for hours. Thank you!

Here's an incomplete snippet of code to give you an idea of what I'm doing...

```
async function getPeople():Promise<PersonList> {
  // This is where I'd like to useSession, but it will blow up if I put it here. Axios works if I use a separate React module
  const res = await fetch( `${process.env.NEXT_PUBLIC_API_URL}/api/contacts`, {
    headers: {
      Authorization: `Bearer 1`,
      'X-Account-Id': '1'
    }
  } );
  return res.json();
}

export default async function People() {
  // This fails when it tries to pre-render...of course, because it's dynamic and shouldn't be pre-rendered
  const people = await getPeople()

  return (
    <>
      <h1 className="text-3xl font-bold py-4">
        Your Community
      </h1>
```

**Comment by jeremyhaile** [comment]
I don't even understand why Next.JS is trying to pre-render this page. The page loads data from an API. It's NOT a static page. Can I just turn off pre-rendering? I just want a page that loads data from an API and then renders.

**Reply by icyJoseph** [reply, marked as answer]
Have you tried using `export const dynamic = 'force-dynamic'` at page level? That should bring you at the `getServerSideProps` level of things, that is, rendered on demand.   

You can also do this per request, IIRC:

```tsx
  const response = await fetch("https://pokeapi.co/api/v2/pokemon/ditto", {
    cache: "no-cache",
  });
```

So that if a path uses this, it'll not be pre-rendered. 

Though to be honest, I thought that a request that uses authorisation headers, was seen as no-cache/no-store. 

Pre-rendering is seen as a good thing, because you could potentially serve parts of the app that do not require runtime data, nor interactivity.

Also, do remember that client components are also run during pre-rendering. 

I also see your app is pointing to its own API route, which is bound to cause you major headaches. Just get the data from where it lives, that's the whole point of Server Components.

Regarding your second question, you can always use the headers function to get data at render time, same with the cookies function, and that should give you enough room to work with. 

Do remember that using axios, server side, would miss out on caching/deduping, however, tools such as `wretch`, which are true fetch wrappers, will work just fine. 

Please do keep on asking, I am also trying to master these new concepts and there's no better way than through answering questions and such.

**Reply by jeremyhaile** [reply]
Thanks so much @icyJoseph - setting `cache: no-cache` on the request did not stop it from pre-rendering (although it's probably a good call anyways)  But setting `export const dynamic = 'force-dynamic'` in the page allowed the build to succeed, my app to deploy, and the page is working. I didn't even realize that was a setting - might be nice if the docs mentioned it near the info on pre-rendering.

I do understand the value of pre-rendering. I just want to be able to control it, and most importantly - disable it when it isn't appropriate.

In regards to using the headers function, the tricky bit is that you can only call useSession from within a react module. Perhaps a good pattern would be to wrap the fetch calls in a React module that calls useSession and uses that to call the headers function. If there is a better pattern to follow, please let me know. 

Thanks again!

**Reply by timbarclay** [reply]
I've found the same thing as @jeremyhaile - the docs say that setting `cache: 'no-cache'` should be enough to trigger a dynamic page, but in my experience, it was still trying to pre-render those pages at build time, which was causing a build failure in my case because my CI server doesn't have access to the API it was trying to call.

Setting `export const dynamic = 'force-dynamic'` appears to be necessary, but that's not what [the docs](https://nextjs.org/docs/app/building-your-application/rendering/static-and-dynamic-rendering#dynamic-data-fetching) say.

So thank you @icyJoseph for pointing me towards something that has fixed my problem, but I'm a bit confused about why it was a problem in the first place.
